mautic/mautic
# Support configuration of RabbitMQ idle timeout

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | N
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | N

#### Description:

Adds the following configuration options to the **QueueBundle**:

- *rabbitmq_idle_timeout*, default `0`.
- *rabbitmq_idle_timeout_exit_code*, default `0`. Doesn't appear to actually get returned by queue commands.

These options are read when loading the `old_sound_rabbit_mq` extension, in the setup for the *mautic* consumer.

##### N.B.

Ideally this would be accompanied by options for the `graceful_max_execution` settings, but there is no way to have an ineffectual default value for that as the extension does an `isset()` test before applying the setting to the service definition (in [OldSoundRabbitMqExtension::loadConsumers][rmqb-loadconsumers]). Therefore any value, like NULL or an empty array, will cause the _graceful max execution time_ to be set. Supporting ommission of null defaults would require changes to the _php-amqplib/rabbitmq-bundle_. As there are no checks against null or zero values, the following issues arise:

* When the `graceful_max_execution` is NULL or `['timeout' => NULL, ...]`, a non-valid DateTime is created [in the consumer][rmqb-consumer-maxexec].
* The use of an actual zero (`0`) for the `['graceful_max_execution']['timeout']` is explicitly [warned against][rmqb-consumer-nozero].

[rmqb-loadconsumers]: https://github.com/php-amqplib/RabbitMqBundle/blob/master/DependencyInjection/OldSoundRabbitMqExtension.php#L216-L225
[rmqb-consumer-maxexec]: https://github.com/php-amqplib/RabbitMqBundle/blob/master/RabbitMq/Consumer.php#L226-L232
[rmqb-consumer-nozero]: https://github.com/php-amqplib/RabbitMqBundle/blob/master/RabbitMq/Consumer.php#L74-L77


#### Steps to test this PR:

1. Set the options in `app/config/local.php`, e.g.:
    ```php
    <?php
    $parameters = array(
        // ...
        'rabbitmq_idle_timeout' => 10,
        'rabbitmq_idle_timeout_exit_code' => 408, // or whatever
        // ...
    );
    ```
3. Clear the app cache.
2. Run a queue command:
    `$ app/console mautic:queue:process -i page_hit`
3. The process should exit after the consumer has idled for the specified time. As tested with an empty queue, the total run time of the command is often ~6 seconds longer than the specified idle timeout (longer when the cache is rebuilt).
